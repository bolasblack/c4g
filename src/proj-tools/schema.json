{
  "$schema": "http://json-schema.org/schema",
  "id": "SchematicsSchematicSchema",
  "title": "Schematic Options Schema",
  "definitions": {
    "includeItem": {
      "$id": "#includeItem",
      "type": "string",
      "enum": [
        "lint-staged",
        "commitlint",
        "githooks",
        "renovate",
        "jest",
        "prettier"
      ]
    }
  },
  "type": "object",
  "properties": {
    "include": {
      "type": "array",
      "items": {
        "$ref": "#includeItem"
      },
      "description": "The project tools want add"
    },
    "cwd": {
      "type": "string",
      "description": "Specify the schematic execute cwd relative to current cwd"
    },
    "interactive": {
      "type": "boolean",
      "default": true,
      "description": "Ask some option interactively"
    },
    "jest-react": {
      "type": "boolean",
      "default": true,
      "description": "Add react support to jest"
    }
  },
  "required": []
}
