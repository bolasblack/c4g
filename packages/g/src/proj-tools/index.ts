import { strings } from '@angular-devkit/core'
import {
  Rule,
  mergeWith,
  chain,
  callRule,
  apply,
  move,
  template,
} from '@angular-devkit/schematics'
import { Observable, of, empty, from } from 'rxjs'
import { mergeMap, toArray } from 'rxjs/operators'
import * as inquirer from 'inquirer'
import * as path from 'path'
import { produce } from 'immer'
import { file as fileSource } from '@c4605/schematic-utils/lib/sources/file'
import {
  installNodePackage,
  NodePackageType,
} from '@c4605/schematic-utils/lib/rules/installNodePackage'
import { file as fileRule } from '@c4605/schematic-utils/lib/rules/file'
import { when } from '@c4605/schematic-utils/lib/rules/when'
import { Schema as Options, IncludeItem } from './schema'

export { featuresEnabled } from './featureEnabled'

export { Options, IncludeItem }

export function main(options: Options): Rule {
  return (tree, ctx) =>
    (options.interactive ? requestOptions(options) : of(options)).pipe(
      mergeMap((opts) => {
        opts.include = opts.include || []
        return from(opts.include)
      }),
      mergeMap(getRule.bind(null, options)),
      toArray(),
      mergeMap((rules) => {
        if (rules.length) {
          rules = rules.concat(
            installNodePackage({
              packageNames: ['@c4605/toolconfs'],
              workingDirectory: options.cwd,
            }),
          )
        }

        return callRule(chain(rules), tree, ctx)
      }),
    )
}

function getRule(options: Options, includeItem: IncludeItem): Observable<Rule> {
  const install = (opts: Parameters<typeof installNodePackage>[0]) =>
    installNodePackage({
      workingDirectory: options.cwd,
      type: NodePackageType.Dev,
      ...opts,
    })

  const filePath = (relativePath: string) =>
    path.join(options.cwd || '', relativePath)

  switch (includeItem) {
    case IncludeItem.Prettier:
      return of(
        chain([
          install({
            packageNames: ['prettier'],
          }),
          fileRule(filePath('package.json'), (content: object) => ({
            ...content,
            prettier: '@c4605/toolconfs/prettierrc',
          })),
        ]),
      )

    case IncludeItem.Commitlint:
      return of(
        chain([
          install({
            packageNames: ['commitlint', '@commitlint/config-conventional'],
          }),
          configFile(options, '.commitlintrc.js'),
        ]),
      )

    case IncludeItem.LintStaged:
      return of(
        chain([
          install({ packageNames: ['lint-staged'] }),
          configFile(options, 'lint-staged.config.js'),
        ]),
      )

    case IncludeItem.Githooks:
      return of(install({ packageNames: ['git-hook-pure'] }))

    case IncludeItem.Renovate:
      return of(configFile(options, 'renovate.json'))

    case IncludeItem.Eslint:
      return of(eslintRule(options))

    case IncludeItem.Jest:
      return of(jestRule(options))
  }

  return empty()
}

function configFile(options: Options, name: string, rules: Rule[] = []): Rule {
  return mergeWith(
    apply(fileSource(`./files/${name}`), [move(options.cwd || ''), ...rules]),
  )
}

export function eslintRule(options: Options): Rule {
  return chain([
    configFile(options, '.eslintrc.json'),
    when(
      () => options.include!.includes(IncludeItem.Prettier),
      fileRule(
        path.join(options.cwd || '', '.eslintrc.json'),
        (content: Record<string, any>) => ({
          ...content,
          extends: [
            ...content.extends,
            './node_modules/@c4605/toolconfs/eslintrc.prettier',
          ],
        }),
      ),
    ),
    installNodePackage({
      packageNames: [
        'eslint',
        options.include!.includes(IncludeItem.Prettier)
          ? 'eslint-config-prettier'
          : null,
        '@typescript-eslint/parser',
        '@typescript-eslint/eslint-plugin',
      ].filter((s): s is string => Boolean(s)),
      workingDirectory: options.cwd,
      type: NodePackageType.Dev,
    }),
  ])
}

export function jestRule(options: Options = {}): Rule {
  const pkgJsonPath = path.join(options.cwd || '', 'package.json')

  const completedOpts: Required<Options> = Object.assign(
    {
      cwd: '',
      include: [],
      interactive: false,
      jestReact: false,
    },
    options,
  )
  const templateOpts = {
    ...completedOpts,
    dasherize: strings.dasherize,
  }

  return chain([
    addJestDependencies(options),
    configFile(options, 'jest.config.js', [template(templateOpts)]),
    when(
      (tree) => tree.exists(pkgJsonPath),
      fileRule(
        pkgJsonPath,
        produce((content: any) => {
          content.scripts = content.scripts || {}
          if (content.scripts.test) {
            content.scripts.test += '&& jest ./src'
          } else {
            content.scripts.test = 'jest ./src'
          }
        }),
      ),
    ),
  ])

  function addJestDependencies(options: Options): Rule {
    const { jestReact } = options

    const deps = ['jest', 'jest-cli', 'ts-jest', '@types/jest']

    if (jestReact) {
      deps.push(
        'enzyme',
        'enzyme-to-json',
        'enzyme-adapter-react-16',
        'identity-obj-proxy',
      )
    }

    return installNodePackage({
      type: NodePackageType.Dev,
      packageNames: deps,
      workingDirectory: options.cwd,
    })
  }
}

export const requestOptions = (
  opts: Partial<Options> = {},
): Observable<Options> =>
  from(
    inquirer.prompt([
      {
        name: 'include',
        type: 'checkbox',
        choices: Object.values(IncludeItem),
        default: opts.include || [],
      },
      {
        name: 'jestReact',
        type: 'confirm',
        message: 'jest add react support?',
        default: false,
        when(ans: Options) {
          return (
            opts.jestReact == null && ans.include!.includes(IncludeItem.Jest)
          )
        },
      },
    ]),
  )
