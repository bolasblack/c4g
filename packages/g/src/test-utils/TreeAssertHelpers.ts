import {
  UnitTestTree,
  SchematicTestRunner,
} from '@angular-devkit/schematics/testing'

export function matchFileStructureSnapshot(
  creator: () => Promise<{ tree: UnitTestTree }>,
) {
  it('match file structure snapshot', async () => {
    const { tree } = await creator()
    expect(tree.files).toMatchSnapshot('file structure')
  })
}

export function matchFileContentSnapshot(
  creator: () => Promise<{ tree: UnitTestTree }>,
) {
  it('match file content snapshot', async () => {
    const { tree } = await creator()
    tree.files.forEach(file => {
      expect(tree.readContent(file)).toMatchSnapshot(`file content: ${file}`)
    })
  })
}

export function matchTasksSnapshot(
  creator: () => Promise<{ runner: SchematicTestRunner }>,
) {
  it('match tasks snapshot', async () => {
    const { runner } = await creator()
    expect(runner.tasks).toMatchSnapshot('rule tasks')
  })
}

export function matchRunResult(
  creator: () => Promise<{
    tree: UnitTestTree
    runner: SchematicTestRunner
  }>,
) {
  matchFileStructureSnapshot(creator)
  matchFileContentSnapshot(creator)
  matchTasksSnapshot(creator)
}

export function assertTreeSnapshot(
  runner: SchematicTestRunner,
  tree: UnitTestTree,
) {
  expect(tree.files).toMatchSnapshot('file structure')

  tree.files.forEach(file => {
    expect(tree.readContent(file)).toMatchSnapshot(`file content: ${file}`)
  })

  expect(runner.tasks).toMatchSnapshot('rule tasks')
}
