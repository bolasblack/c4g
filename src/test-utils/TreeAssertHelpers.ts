import { UnitTestTree } from '@angular-devkit/schematics/testing'
import { SchematicTestRunner } from '../test-utils/SchematicTestRunner'

export function assertTreeSnapshot(
  runner: SchematicTestRunner,
  tree: UnitTestTree,
) {
  // @ts-ignore
  expect(tree.files).toMatchSnapshot('file structure')
  tree.files.forEach(file => {
    // @ts-ignore
    expect(tree.readContent(file)).toMatchSnapshot(`file content: ${file}`)
  })

  // @ts-ignore
  expect(runner.tasks).toMatchSnapshot('rule tasks')
}
