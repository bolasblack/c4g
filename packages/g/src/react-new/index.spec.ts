import * as path from 'path'
import { Tree, callRule } from '@angular-devkit/schematics'
import { SchematicTestRunner } from '@c4605/schematic-utils/lib/testing/SchematicTestRunner'
import { assertTreeSnapshot } from '../test-utils/TreeAssertHelpers'
import { Options, main as reactNewRuleFactory } from './index'
import { FeaturesEnabledTypedSchematicContext } from '../proj-tools/featureEnabled'
import { IncludeItem } from '../proj-tools'

const collectionPath = path.join(__dirname, '../collection.json')

describe('react-new', () => {
  it('works', async () => {
    const runner = new SchematicTestRunner('schematics', collectionPath)
    const resultTree = await runner
      .runSchematicAsync<Options>(
        'react-new',
        { name: 'hello', interactive: false },
        Tree.empty(),
      )
      .toPromise()
    assertTreeSnapshot(runner, resultTree)
  })

  it('support `author` option', async () => {
    const runner = new SchematicTestRunner('schematics', collectionPath)
    const tree = await runner
      .runSchematicAsync<Options>(
        'react-new',
        {
          name: 'hello',
          author: 'c4605 <bolasblack@gmail.com>',
          interactive: false,
        },
        Tree.empty(),
      )
      .toPromise()
    assertTreeSnapshot(runner, tree)
  })

  it('add spec files if user added proj-tool jest', async () => {
    const runner = new SchematicTestRunner('schematics', collectionPath)

    const tree = await runner
      .runSchematicAsync(
        'react-new',
        {
          name: 'hello',
          author: 'c4605 <bolasblack@gmail.com>',
          interactive: false,
        },
        undefined,
        (schematic, opts, tree, oriCtx) => {
          const ctx: FeaturesEnabledTypedSchematicContext = runner.createContext(
            'react-new',
            oriCtx,
          )
          ctx.enabledFeatures = [IncludeItem.Jest]
          return callRule(reactNewRuleFactory(opts), tree, ctx)
        },
      )
      .toPromise()

    assertTreeSnapshot(runner, tree)
  })
})
