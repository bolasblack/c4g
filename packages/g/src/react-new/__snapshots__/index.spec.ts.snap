// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`react-new add spec files if user added proj-tool jest: file content: /hello/.babelrc.js 1`] = `
"const path = require('path')

module.exports = {
  plugins: [
    [
      // astroturf/css-loader not compatible with thread-loader, so we use babel plugin
      'astroturf/plugin',
      {
        tagName: 'css',
        extension: '.scss',
        writeFiles: true, // Writes css files to disk using the result of \`getFileName\`
        getFileName(hostFilePath, pluginsOptions) {
          const basepath = path.join(
            path.dirname(hostFilePath),
            path.basename(hostFilePath, path.extname(hostFilePath)),
          )
          const relativePath = path.relative(__dirname, basepath)
          return \`.astroturf/extracted_styles/\${relativePath}.scss\`
        },
      },
    ],
  ],
}
"
`;

exports[`react-new add spec files if user added proj-tool jest: file content: /hello/.eslintrc.json 1`] = `
"{
  \\"extends\\": [
    \\"./node_modules/@c4605/toolconfs/eslintrc.base\\",
    \\"./node_modules/@c4605/toolconfs/eslintrc.ts\\",
    \\"./node_modules/@c4605/toolconfs/eslintrc.prettier\\",
    \\"./node_modules/@c4605/toolconfs/eslintrc.react\\"
  ],
  \\"parserOptions\\": {
    \\"project\\": \\"./tsconfig.json\\"
  },
  \\"rules\\": {}
}"
`;

exports[`react-new add spec files if user added proj-tool jest: file content: /hello/.gitignore 1`] = `
"# Logs
logs
*.log
npm-debug.log*
yarn-debug.log*
yarn-error.log*
lerna-debug.log*

# Editors
# vscode
.vscode
# intelj
.idea
# vim
*.swp
*~
# emacs
\\\\#*\\\\#
.\\\\#*

# Diagnostic reports (https://nodejs.org/api/report.html)
report.[0-9]*.[0-9]*.[0-9]*.[0-9]*.json

# Runtime data
pids
*.pid
*.seed
*.pid.lock

# Directory for instrumented libs generated by jscoverage/JSCover
lib-cov

# Coverage directory used by tools like istanbul
coverage
*.lcov

# nyc test coverage
.nyc_output

# Grunt intermediate storage (https://gruntjs.com/creating-plugins#storing-task-files)
.grunt

# Bower dependency directory (https://bower.io/)
bower_components

# node-waf configuration
.lock-wscript

# Compiled binary addons (https://nodejs.org/api/addons.html)
build/Release

# Dependency directories
node_modules/
jspm_packages/

# TypeScript v1 declaration files
typings/

# TypeScript cache
*.tsbuildinfo

# Optional npm cache directory
.npm

# Optional eslint cache
.eslintcache

# Optional REPL history
.node_repl_history

# Output of 'npm pack'
*.tgz

# Yarn Integrity file
.yarn-integrity

# dotenv environment variables file
.env
.env.*
.envrc
.envrc.*

# parcel-bundler cache (https://parceljs.org/)
.cache

# next.js build output
.next

# nuxt.js build output
.nuxt

# rollup.js default build output
dist/

# Serverless directories
.serverless/

# Temporary folders
tmp/
temp/

# Misc
.DS_*
"
`;

exports[`react-new add spec files if user added proj-tool jest: file content: /hello/README.md 1`] = `
"# hello

## Usage

\`\`\`bash
yarn start # watch and serve code
yarn build # build code
\`\`\`
"
`;

exports[`react-new add spec files if user added proj-tool jest: file content: /hello/package.json 1`] = `
"{
  \\"name\\": \\"hello\\",
  \\"version\\": \\"0.1.0\\",
  \\"main\\": null,
  \\"author\\": \\"c4605 <bolasblack@gmail.com>\\",
  \\"license\\": \\"MIT\\",
  \\"scripts\\": {
    \\"start\\": \\"poi -r ts-node/register --parallel --serve ./src\\",
    \\"build\\": \\"poi -r ts-node/register --parallel ./src\\"
  },
  \\"devDependencies\\": {
    \\"prettier\\": \\"^1.0.0\\",
    \\"eslint\\": \\"^1.0.0\\",
    \\"eslint-config-prettier\\": \\"^1.0.0\\",
    \\"@typescript-eslint/parser\\": \\"^1.0.0\\",
    \\"@typescript-eslint/eslint-plugin\\": \\"^1.0.0\\",
    \\"typescript\\": \\"^1.0.0\\",
    \\"ts-node\\": \\"^1.0.0\\",
    \\"@types/node\\": \\"^1.0.0\\",
    \\"poi\\": \\"^1.0.0\\",
    \\"@types/poi\\": \\"^1.0.0\\",
    \\"@poi/plugin-typescript\\": \\"^1.0.0\\",
    \\"sass\\": \\"^1.0.0\\",
    \\"sass-loader\\": \\"^1.0.0\\",
    \\"react-hot-loader\\": \\"^1.0.0\\"
  },
  \\"dependencies\\": {
    \\"@c4605/toolconfs\\": \\"^1.0.0\\",
    \\"react\\": \\"^1.0.0\\",
    \\"@types/react\\": \\"^1.0.0\\",
    \\"react-dom\\": \\"^1.0.0\\",
    \\"@types/react-dom\\": \\"^1.0.0\\",
    \\"astroturf\\": \\"^1.0.0\\"
  },
  \\"prettier\\": \\"@c4605/toolconfs/prettierrc\\"
}"
`;

exports[`react-new add spec files if user added proj-tool jest: file content: /hello/poi.config.ts 1`] = `
"import { Config } from 'poi'

const config: Config = {
  entry: 'src/index.tsx',
  plugins: [
    {
      resolve: '@poi/plugin-typescript',
      options: {
        babel: true,
      },
    },
  ],
}

export default config
"
`;

exports[`react-new add spec files if user added proj-tool jest: file content: /hello/src/index.tsx 1`] = `
"import 'react-hot-loader'
import * as React from 'react'
import * as ReactDOM from 'react-dom'
import { css } from 'astroturf'

function renderApp() {
  ReactDOM.render(<h1>It works</h1>, document.getElementById('app')!)
}

renderApp()
;(module as any).hot.accept(renderApp)

css\`
  h1 {
    font-size: 32px;
  }
\`
"
`;

exports[`react-new add spec files if user added proj-tool jest: file content: /hello/tsconfig.json 1`] = `
"{
  \\"extends\\": \\"@c4605/toolconfs/tsconfig\\",
  \\"compilerOptions\\": {
    \\"jsx\\": \\"react\\"
  }
}"
`;

exports[`react-new add spec files if user added proj-tool jest: file structure 1`] = `
Array [
  "/hello/package.json",
  "/hello/.eslintrc.json",
  "/hello/.babelrc.js",
  "/hello/README.md",
  "/hello/poi.config.ts",
  "/hello/tsconfig.json",
  "/hello/.gitignore",
  "/hello/src/index.tsx",
]
`;

exports[`react-new add spec files if user added proj-tool jest: rule tasks 1`] = `
Array [
  Object {
    "name": "@c4g/shell-exec-task",
    "options": Object {
      "args": Array [],
      "command": "yarn",
      "options": Object {
        "cwd": "hello",
      },
    },
  },
  Object {
    "name": "@c4g/shell-exec-task",
    "options": Object {
      "args": Array [],
      "command": "yarn",
      "options": Object {
        "cwd": "hello",
      },
    },
  },
  Object {
    "name": "@c4g/shell-exec-task",
    "options": Object {
      "args": Array [],
      "command": "yarn",
      "options": Object {
        "cwd": "hello",
      },
    },
  },
  Object {
    "name": "@c4g/shell-exec-task",
    "options": Object {
      "args": Array [],
      "command": "yarn",
      "options": Object {
        "cwd": "hello",
      },
    },
  },
  Object {
    "name": "@c4g/shell-exec-task",
    "options": Object {
      "args": Array [],
      "command": "yarn",
      "options": Object {
        "cwd": "hello",
      },
    },
  },
]
`;

exports[`react-new support \`author\` option: file content: /hello/.babelrc.js 1`] = `
"const path = require('path')

module.exports = {
  plugins: [
    [
      // astroturf/css-loader not compatible with thread-loader, so we use babel plugin
      'astroturf/plugin',
      {
        tagName: 'css',
        extension: '.scss',
        writeFiles: true, // Writes css files to disk using the result of \`getFileName\`
        getFileName(hostFilePath, pluginsOptions) {
          const basepath = path.join(
            path.dirname(hostFilePath),
            path.basename(hostFilePath, path.extname(hostFilePath)),
          )
          const relativePath = path.relative(__dirname, basepath)
          return \`.astroturf/extracted_styles/\${relativePath}.scss\`
        },
      },
    ],
  ],
}
"
`;

exports[`react-new support \`author\` option: file content: /hello/.eslintrc.json 1`] = `
"{
  \\"extends\\": [
    \\"./node_modules/@c4605/toolconfs/eslintrc.base\\",
    \\"./node_modules/@c4605/toolconfs/eslintrc.ts\\",
    \\"./node_modules/@c4605/toolconfs/eslintrc.prettier\\",
    \\"./node_modules/@c4605/toolconfs/eslintrc.react\\"
  ],
  \\"parserOptions\\": {
    \\"project\\": \\"./tsconfig.json\\"
  },
  \\"rules\\": {}
}"
`;

exports[`react-new support \`author\` option: file content: /hello/.gitignore 1`] = `
"# Logs
logs
*.log
npm-debug.log*
yarn-debug.log*
yarn-error.log*
lerna-debug.log*

# Editors
# vscode
.vscode
# intelj
.idea
# vim
*.swp
*~
# emacs
\\\\#*\\\\#
.\\\\#*

# Diagnostic reports (https://nodejs.org/api/report.html)
report.[0-9]*.[0-9]*.[0-9]*.[0-9]*.json

# Runtime data
pids
*.pid
*.seed
*.pid.lock

# Directory for instrumented libs generated by jscoverage/JSCover
lib-cov

# Coverage directory used by tools like istanbul
coverage
*.lcov

# nyc test coverage
.nyc_output

# Grunt intermediate storage (https://gruntjs.com/creating-plugins#storing-task-files)
.grunt

# Bower dependency directory (https://bower.io/)
bower_components

# node-waf configuration
.lock-wscript

# Compiled binary addons (https://nodejs.org/api/addons.html)
build/Release

# Dependency directories
node_modules/
jspm_packages/

# TypeScript v1 declaration files
typings/

# TypeScript cache
*.tsbuildinfo

# Optional npm cache directory
.npm

# Optional eslint cache
.eslintcache

# Optional REPL history
.node_repl_history

# Output of 'npm pack'
*.tgz

# Yarn Integrity file
.yarn-integrity

# dotenv environment variables file
.env
.env.*
.envrc
.envrc.*

# parcel-bundler cache (https://parceljs.org/)
.cache

# next.js build output
.next

# nuxt.js build output
.nuxt

# rollup.js default build output
dist/

# Serverless directories
.serverless/

# Temporary folders
tmp/
temp/

# Misc
.DS_*
"
`;

exports[`react-new support \`author\` option: file content: /hello/README.md 1`] = `
"# hello

## Usage

\`\`\`bash
yarn start # watch and serve code
yarn build # build code
\`\`\`
"
`;

exports[`react-new support \`author\` option: file content: /hello/package.json 1`] = `
"{
  \\"name\\": \\"hello\\",
  \\"version\\": \\"0.1.0\\",
  \\"main\\": null,
  \\"author\\": \\"c4605 <bolasblack@gmail.com>\\",
  \\"license\\": \\"MIT\\",
  \\"scripts\\": {
    \\"start\\": \\"poi -r ts-node/register --parallel --serve ./src\\",
    \\"build\\": \\"poi -r ts-node/register --parallel ./src\\"
  },
  \\"devDependencies\\": {
    \\"prettier\\": \\"^1.0.0\\",
    \\"eslint\\": \\"^1.0.0\\",
    \\"eslint-config-prettier\\": \\"^1.0.0\\",
    \\"@typescript-eslint/parser\\": \\"^1.0.0\\",
    \\"@typescript-eslint/eslint-plugin\\": \\"^1.0.0\\",
    \\"typescript\\": \\"^1.0.0\\",
    \\"ts-node\\": \\"^1.0.0\\",
    \\"@types/node\\": \\"^1.0.0\\",
    \\"poi\\": \\"^1.0.0\\",
    \\"@types/poi\\": \\"^1.0.0\\",
    \\"@poi/plugin-typescript\\": \\"^1.0.0\\",
    \\"sass\\": \\"^1.0.0\\",
    \\"sass-loader\\": \\"^1.0.0\\",
    \\"react-hot-loader\\": \\"^1.0.0\\"
  },
  \\"dependencies\\": {
    \\"@c4605/toolconfs\\": \\"^1.0.0\\",
    \\"react\\": \\"^1.0.0\\",
    \\"@types/react\\": \\"^1.0.0\\",
    \\"react-dom\\": \\"^1.0.0\\",
    \\"@types/react-dom\\": \\"^1.0.0\\",
    \\"astroturf\\": \\"^1.0.0\\"
  },
  \\"prettier\\": \\"@c4605/toolconfs/prettierrc\\"
}"
`;

exports[`react-new support \`author\` option: file content: /hello/poi.config.ts 1`] = `
"import { Config } from 'poi'

const config: Config = {
  entry: 'src/index.tsx',
  plugins: [
    {
      resolve: '@poi/plugin-typescript',
      options: {
        babel: true,
      },
    },
  ],
}

export default config
"
`;

exports[`react-new support \`author\` option: file content: /hello/src/index.tsx 1`] = `
"import 'react-hot-loader'
import * as React from 'react'
import * as ReactDOM from 'react-dom'
import { css } from 'astroturf'

function renderApp() {
  ReactDOM.render(<h1>It works</h1>, document.getElementById('app')!)
}

renderApp()
;(module as any).hot.accept(renderApp)

css\`
  h1 {
    font-size: 32px;
  }
\`
"
`;

exports[`react-new support \`author\` option: file content: /hello/tsconfig.json 1`] = `
"{
  \\"extends\\": \\"@c4605/toolconfs/tsconfig\\",
  \\"compilerOptions\\": {
    \\"jsx\\": \\"react\\"
  }
}"
`;

exports[`react-new support \`author\` option: file structure 1`] = `
Array [
  "/hello/package.json",
  "/hello/.eslintrc.json",
  "/hello/.babelrc.js",
  "/hello/README.md",
  "/hello/poi.config.ts",
  "/hello/tsconfig.json",
  "/hello/.gitignore",
  "/hello/src/index.tsx",
]
`;

exports[`react-new support \`author\` option: rule tasks 1`] = `
Array [
  Object {
    "name": "@c4g/shell-exec-task",
    "options": Object {
      "args": Array [],
      "command": "yarn",
      "options": Object {
        "cwd": "hello",
      },
    },
  },
  Object {
    "name": "@c4g/shell-exec-task",
    "options": Object {
      "args": Array [],
      "command": "yarn",
      "options": Object {
        "cwd": "hello",
      },
    },
  },
  Object {
    "name": "@c4g/shell-exec-task",
    "options": Object {
      "args": Array [],
      "command": "yarn",
      "options": Object {
        "cwd": "hello",
      },
    },
  },
  Object {
    "name": "@c4g/shell-exec-task",
    "options": Object {
      "args": Array [],
      "command": "yarn",
      "options": Object {
        "cwd": "hello",
      },
    },
  },
  Object {
    "name": "@c4g/shell-exec-task",
    "options": Object {
      "args": Array [],
      "command": "yarn",
      "options": Object {
        "cwd": "hello",
      },
    },
  },
]
`;

exports[`react-new works: file content: /hello/.babelrc.js 1`] = `
"const path = require('path')

module.exports = {
  plugins: [
    [
      // astroturf/css-loader not compatible with thread-loader, so we use babel plugin
      'astroturf/plugin',
      {
        tagName: 'css',
        extension: '.scss',
        writeFiles: true, // Writes css files to disk using the result of \`getFileName\`
        getFileName(hostFilePath, pluginsOptions) {
          const basepath = path.join(
            path.dirname(hostFilePath),
            path.basename(hostFilePath, path.extname(hostFilePath)),
          )
          const relativePath = path.relative(__dirname, basepath)
          return \`.astroturf/extracted_styles/\${relativePath}.scss\`
        },
      },
    ],
  ],
}
"
`;

exports[`react-new works: file content: /hello/.eslintrc.json 1`] = `
"{
  \\"extends\\": [
    \\"./node_modules/@c4605/toolconfs/eslintrc.base\\",
    \\"./node_modules/@c4605/toolconfs/eslintrc.ts\\",
    \\"./node_modules/@c4605/toolconfs/eslintrc.prettier\\",
    \\"./node_modules/@c4605/toolconfs/eslintrc.react\\"
  ],
  \\"parserOptions\\": {
    \\"project\\": \\"./tsconfig.json\\"
  },
  \\"rules\\": {}
}"
`;

exports[`react-new works: file content: /hello/.gitignore 1`] = `
"# Logs
logs
*.log
npm-debug.log*
yarn-debug.log*
yarn-error.log*
lerna-debug.log*

# Editors
# vscode
.vscode
# intelj
.idea
# vim
*.swp
*~
# emacs
\\\\#*\\\\#
.\\\\#*

# Diagnostic reports (https://nodejs.org/api/report.html)
report.[0-9]*.[0-9]*.[0-9]*.[0-9]*.json

# Runtime data
pids
*.pid
*.seed
*.pid.lock

# Directory for instrumented libs generated by jscoverage/JSCover
lib-cov

# Coverage directory used by tools like istanbul
coverage
*.lcov

# nyc test coverage
.nyc_output

# Grunt intermediate storage (https://gruntjs.com/creating-plugins#storing-task-files)
.grunt

# Bower dependency directory (https://bower.io/)
bower_components

# node-waf configuration
.lock-wscript

# Compiled binary addons (https://nodejs.org/api/addons.html)
build/Release

# Dependency directories
node_modules/
jspm_packages/

# TypeScript v1 declaration files
typings/

# TypeScript cache
*.tsbuildinfo

# Optional npm cache directory
.npm

# Optional eslint cache
.eslintcache

# Optional REPL history
.node_repl_history

# Output of 'npm pack'
*.tgz

# Yarn Integrity file
.yarn-integrity

# dotenv environment variables file
.env
.env.*
.envrc
.envrc.*

# parcel-bundler cache (https://parceljs.org/)
.cache

# next.js build output
.next

# nuxt.js build output
.nuxt

# rollup.js default build output
dist/

# Serverless directories
.serverless/

# Temporary folders
tmp/
temp/

# Misc
.DS_*
"
`;

exports[`react-new works: file content: /hello/README.md 1`] = `
"# hello

## Usage

\`\`\`bash
yarn start # watch and serve code
yarn build # build code
\`\`\`
"
`;

exports[`react-new works: file content: /hello/package.json 1`] = `
"{
  \\"name\\": \\"hello\\",
  \\"version\\": \\"0.1.0\\",
  \\"main\\": null,
  \\"license\\": \\"MIT\\",
  \\"scripts\\": {
    \\"start\\": \\"poi -r ts-node/register --parallel --serve ./src\\",
    \\"build\\": \\"poi -r ts-node/register --parallel ./src\\"
  },
  \\"devDependencies\\": {
    \\"prettier\\": \\"^1.0.0\\",
    \\"eslint\\": \\"^1.0.0\\",
    \\"eslint-config-prettier\\": \\"^1.0.0\\",
    \\"@typescript-eslint/parser\\": \\"^1.0.0\\",
    \\"@typescript-eslint/eslint-plugin\\": \\"^1.0.0\\",
    \\"typescript\\": \\"^1.0.0\\",
    \\"ts-node\\": \\"^1.0.0\\",
    \\"@types/node\\": \\"^1.0.0\\",
    \\"poi\\": \\"^1.0.0\\",
    \\"@types/poi\\": \\"^1.0.0\\",
    \\"@poi/plugin-typescript\\": \\"^1.0.0\\",
    \\"sass\\": \\"^1.0.0\\",
    \\"sass-loader\\": \\"^1.0.0\\",
    \\"react-hot-loader\\": \\"^1.0.0\\"
  },
  \\"dependencies\\": {
    \\"@c4605/toolconfs\\": \\"^1.0.0\\",
    \\"react\\": \\"^1.0.0\\",
    \\"@types/react\\": \\"^1.0.0\\",
    \\"react-dom\\": \\"^1.0.0\\",
    \\"@types/react-dom\\": \\"^1.0.0\\",
    \\"astroturf\\": \\"^1.0.0\\"
  },
  \\"prettier\\": \\"@c4605/toolconfs/prettierrc\\"
}"
`;

exports[`react-new works: file content: /hello/poi.config.ts 1`] = `
"import { Config } from 'poi'

const config: Config = {
  entry: 'src/index.tsx',
  plugins: [
    {
      resolve: '@poi/plugin-typescript',
      options: {
        babel: true,
      },
    },
  ],
}

export default config
"
`;

exports[`react-new works: file content: /hello/src/index.tsx 1`] = `
"import 'react-hot-loader'
import * as React from 'react'
import * as ReactDOM from 'react-dom'
import { css } from 'astroturf'

function renderApp() {
  ReactDOM.render(<h1>It works</h1>, document.getElementById('app')!)
}

renderApp()
;(module as any).hot.accept(renderApp)

css\`
  h1 {
    font-size: 32px;
  }
\`
"
`;

exports[`react-new works: file content: /hello/tsconfig.json 1`] = `
"{
  \\"extends\\": \\"@c4605/toolconfs/tsconfig\\",
  \\"compilerOptions\\": {
    \\"jsx\\": \\"react\\"
  }
}"
`;

exports[`react-new works: file structure 1`] = `
Array [
  "/hello/package.json",
  "/hello/.eslintrc.json",
  "/hello/.babelrc.js",
  "/hello/README.md",
  "/hello/poi.config.ts",
  "/hello/tsconfig.json",
  "/hello/.gitignore",
  "/hello/src/index.tsx",
]
`;

exports[`react-new works: rule tasks 1`] = `
Array [
  Object {
    "name": "@c4g/shell-exec-task",
    "options": Object {
      "args": Array [],
      "command": "yarn",
      "options": Object {
        "cwd": "hello",
      },
    },
  },
  Object {
    "name": "@c4g/shell-exec-task",
    "options": Object {
      "args": Array [],
      "command": "yarn",
      "options": Object {
        "cwd": "hello",
      },
    },
  },
  Object {
    "name": "@c4g/shell-exec-task",
    "options": Object {
      "args": Array [],
      "command": "yarn",
      "options": Object {
        "cwd": "hello",
      },
    },
  },
  Object {
    "name": "@c4g/shell-exec-task",
    "options": Object {
      "args": Array [],
      "command": "yarn",
      "options": Object {
        "cwd": "hello",
      },
    },
  },
  Object {
    "name": "@c4g/shell-exec-task",
    "options": Object {
      "args": Array [],
      "command": "yarn",
      "options": Object {
        "cwd": "hello",
      },
    },
  },
]
`;
